common/flatpak-instance.c

```diff
@@ -0,0 +1,317 @@
+get_pid (const char *dir)
+{
+  g_autofree char *file = NULL;
+  g_autofree char *contents = NULL;
+  g_autoptr(GError) error = NULL;
+
+  file = g_build_filename (dir, "pid", NULL);
+
+  if (!g_file_get_contents (file, &contents, NULL, &error))
+    {
+      g_debug ("Failed to load pid file '%s': %s", file, error->message);
+      return 0;
+    }
+
+  return (int) g_ascii_strtoll (contents, NULL, 10);
+}
+
+static FlatpakInstance *
+flatpak_instance_new (const char *id)
+{
+  FlatpakInstance *self = g_object_new (flatpak_instance_get_type (), NULL);
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  priv->id = g_strdup (id);
+
+  priv->dir = g_build_filename (g_get_user_runtime_dir (), ".flatpak", id, NULL);
+
+  priv->pid = get_pid (priv->dir);
+  priv->child_pid = get_child_pid (priv->dir);
+  priv->info = get_instance_info (priv->dir);
+
+  if (priv->info)
+    {
+      priv->app = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_APPLICATION, FLATPAK_METADATA_KEY_NAME, NULL);
+      priv->runtime = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_APPLICATION, FLATPAK_METADATA_KEY_RUNTIME, NULL);
+
+      priv->arch = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_INSTANCE, FLATPAK_METADATA_KEY_ARCH, NULL);
+      priv->branch = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_INSTANCE, FLATPAK_METADATA_KEY_BRANCH, NULL);
+      priv->commit = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_INSTANCE, FLATPAK_METADATA_KEY_APP_COMMIT, NULL);
+      priv->runtime_commit = g_key_file_get_string (priv->info,
+          FLATPAK_METADATA_GROUP_INSTANCE, FLATPAK_METADATA_KEY_RUNTIME_COMMIT, NULL);
+    }
+
+  return self;
+}
+
+GPtrArray *
+flatpak_instance_get_all (void)
+{
+  g_autoptr(GPtrArray) instances = NULL;
+  g_autofree char *base_dir = NULL;
+  g_autoptr(GFile) file = NULL;
+  GFileInfo *info;
+  g_autoptr(GFileEnumerator) enumerator = NULL;
+  g_autoptr(GError) error = NULL;
+
+  flatpak_run_gc_ids ();
+
+  instances = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+  
+  base_dir = g_build_filename (g_get_user_runtime_dir (), ".flatpak", NULL);
+  file = g_file_new_for_path (base_dir);
+  enumerator = g_file_enumerate_children (file,
+                                          G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                          G_FILE_QUERY_INFO_NONE,
+                                          NULL,
+                                          &error);
+
+  if (enumerator == NULL)
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          /* nothing to show */
+          return g_steal_pointer (&instances);
+        }
+
+      return NULL;
+    }
+
+  while ((info = g_file_enumerator_next_file (enumerator, NULL, &error)) != NULL)
+    {
+      g_ptr_array_add (instances, flatpak_instance_new (g_file_info_get_name (info)));
```

Comment:
This should also do the lock thing to not return instances for a stale app.

Reply:
I'm calling the gc thing beforehand, is that not good enough ?